File: k_pair.py
'''
TC:O(NLogN)
SC:O(N)
'''

class Solution:
    def findPairs(self, nums, k) -> int:
        nums = sorted(nums)

        left = 0
        right = 1

        result = 0

        while (left < len(nums) and right < len(nums)):
            if (left == right or nums[right] - nums[left] < k):
                right += 1
            elif nums[right] - nums[left] > k:
                left += 1
            else:
                left += 1
                result += 1
                while (left < len(nums) and nums[left] == nums[left - 1]):
                    left += 1

        return result

print(Solution().findPairs([-1,-2,-3],1))

File: kPair.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

class Solution {
    public int findPairs(int[] nums, int k) {
        if(nums.length==0 || nums==null)
            return 0;
        
        int count=0;
        Map<Integer, Integer> hMap = new HashMap<>();
        
        //updating values in hashmap
        for(int i: nums) {
            hMap.put(i, hMap.getOrDefault(i,0)+1);
        }
        
        //update the count
        for(int i: hMap.keySet()) {
            if(k==0) {
                if(hMap.get(i)>=2)
                    count++;
            }
            else {
                if(hMap.containsKey(i+k))
                    count++;
            }
        }
        return count;
    }
}